### ISSUE TITLE
Assert raises test failure for test_ill_condition_warning

### ISSUE THREAD

**Issue by matthew-brett** [scipy/scipy#6984]
Testing against numpy 1.8.2, I get the following test error for master:

```
ERROR: test_ill_condition_warning (test_basic.TestSolve)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/venv/local/lib/python2.7/site-packages/scipy/linalg/tests/test_basic.py", line 714, in test_ill_condition_warning
    assert_raises(RuntimeWarning, solve, a, b)
  File "/venv/local/lib/python2.7/site-packages/numpy/testing/utils.py", line 1020, in assert_raises
    return nose.tools.assert_raises(*args,**kwargs)
  File "/usr/lib/python2.7/unittest/case.py", line 475, in assertRaises
    callableObj(*args, **kwargs)
  File "/venv/local/lib/python2.7/site-packages/scipy/linalg/basic.py", line 219, in solve
    raise LinAlgError('Matrix is singular.')
LinAlgError: Matrix is singular.
```

https://travis-ci.org/MacPython/scipy-wheels/jobs/195612819#L488

Should `LinAlgError` also be allowed for this case?

**Comment by ilayn** [comment]
Strange. That shouldn't be marked as singular as I've tested in various places, except, of course, OSX variants.

**Comment by matthew-brett** [comment]
Have you tested with the current numpy 1.8.2 wheel?

**Comment by ilayn** [comment]
You mean numpy 1.12?

**Comment by matthew-brett** [comment]
No, I meant 1.8.2 - the current scipy minimum dependency.

**Comment by ilayn** [comment]
Ah true, no I have not. Actually I've always built the sources. That is a good reminder for me. 

I'm still curious about the reason though. SciPy should be independent from the NumPy versions on this matter as it directly calls ?GESV. Maybe get_lapack_func does something strange.

**Comment by matthew-brett** [comment]
By the way - the test failures are for Linux 64-bit in Ubuntu trusty containers.  Weirdly, 32-bit passes.  See : https://travis-ci.org/MacPython/scipy-wheels/builds/195612818

**Comment by matthew-brett** [comment]
Also - is it possible you are not building against OpenBLAS, as I am for the wheel builds?

**Comment by ilayn** [comment]
No, I'm not using OpenBLAS at all. Pretty much off-the-shelf virtual environment with Ubuntu 16.04 installation. Darn, this is indeed a very delicate business, I should try to automate these builds.

**Comment by matthew-brett** [comment]
Hum - yes - tricky.  The openblas / wheel case will at least be tested in master, because I've set up a daily cron job to build scipy wheels - which is how I found this guy.

**Comment by ilayn** [comment]
I still couldn't replicate this. Could you please comment out that test and run 

    a = np.range(1, 10).reshape(3, 3)
    b = np.ones(3)*15
    _, _, _, _, _, _, _, x, rcond, _, _, info = sp.linalg.lapack.dgesvx(a, b)

and check the resulting condition number ? Or if `info ==  4`? You can try with `sgesvx` too just to check.

**Comment by matthew-brett** [comment]
Anything else I can do to help here?

**Comment by ilayn** [comment]
I still have some difficulty getting my virtual environment recognizing OpenBLAS so maybe the easiest way is to change the test to something that would emit a warning in both settings. 

Could you please check whether 

```
a = np.array([[1, 1], [1+1e-16, 1-1e-16]]) 
b = np.ones(2)
sp.linalg.solve(a, b)
```
emits the warning? 

If so I'll modify the test since we only need the warning not the solution precision. Thus, anything goes in this particular case.

**Comment by matthew-brett** [comment]
The way I got the scipy compile working with openblas, was to unpack https://3f23b170c54c2533c070-1c8a9b3114517dc5fe17b7c3f8c63a43.ssl.cf2.rackcdn.com/openblas-0.2.18-Linux-x86_64.tar.gz into `/home/mb312/tmp` and compile scipy with the following `site.cfg` in my `scipy` root directory:

```
[openblas]
libraries = openblas
library_dirs = /home/mb312/tmp/usr/local/lib
include_dirs = /home/mb312/tmp/usr/local/include
runtime_library_dirs = /home/mb312/tmp/usr/local/lib
```

**Comment by ilayn** [comment]
Thanks, I think I did a similar thing but maybe I screwed up somewhere. I hope I can try again this evening again with this config.

**Comment by matthew-brett** [comment]
This code (from your comment above):

```
a = np.array([[1, 1], [1+1e-16, 1-1e-16]]) 
b = np.ones(2)
sp.linalg.solve(a, b)
```

gives this:

```
/home/mb312/dev_trees/scipy/scipy/linalg/basic.py:223: RuntimeWarning: scipy.linalg.solve
Ill-conditioned matrix detected. Result is not guaranteed to be accurate.
Reciprocal condition number: 2.77555756156e-17
  ' condition number: {}'.format(rcond), RuntimeWarning)
Out[7]: array([ 1., -0.])
```

**Comment by ilayn** [comment]
Yes, that was the warning the test was looking for but failed to find due to the singular matrix error.

**Comment by matthew-brett** [comment]
OK to just go for that change to the test then?

**Comment by pv** [comment]
Regardless of whether you can reproduce it, the problem is in raising 
LinAlgError for zero pivot. Rounding error can easily make the pivot 
eps-size. So you probably need to change the test to  accept that behavior.
.
Of course, the raise-on-zero-pivot behavior is sort of misleading 
nonsense, as badly conditioned solutions may be completely wrong in any 
case, so not getting a LinAlgError does not guarantee the solution is 
sensible. This apparently tends to mislead people, and then cause 
surprises when a matrix that is singular in exact arithmetic is 
successfulyl inverted. (See the many discussions on numpy issue tracker 
on the topic of "numpy inverts a singular matrix?!")
.
Other computation environments (octave, matlab) return a matrix full of 
infs if there's a zero pivot + emit a warning.
.
Changing the behavior in numpy/scipy to something more sensible was 
discussed, but the backward compat issues are not fully sorted out yet.

**Comment by ilayn** [comment]
@pv Here it is the other way around, when there is ill-conditioning matlab emits a warning and returns something (right or wrong, we use the same solver so no discrepancy here). Whenever LAPACK finds a conditioning number too high it emits the zero-pivot flag and we raise LinalgError so rounding is not done in matlab or scipy highlevel it is decided by LAPACK. 

Here I'm just trying to test whether the warning mechanism works or not. Your points are valid but that's the reason we swithced to expert version of DGESV. Apparently, emitting the warning safely is the culprit as BLAS versions make a difference in internal manipulations.

**Comment by pv** [comment]
I think the statement that "LAPACK finds a conditioning number too high 
it emits the zero-pivot flag" is not correct. The zero-pivot flag is 
emitted if the pivot is really exactly zero --- whether this happens 
depends on how the rounding error happens to go in the algorithm.
.
Since you cannot control that across platforms, either the test matrix 
needs to be less singular, or the test should accept also LinAlgError.

**Comment by ilayn** [comment]
Yes, let me rephrase it as there is no other decision mechanism other than LAPACK's DLAMCH stuff to find an exact zero. 

What I meant is that for this particular test, I wanted to test specifically the ill-conditioning warning without the exception. Less singular example apparently is tricky since some other user might come with another example that fails for them too. I guess accepting the exception seems to be the logical choice. 

For others here is the relevant issue that was mentioned before https://github.com/numpy/numpy/issues/2074

**Comment by pv** [comment]
AFAIK there's actually no DLAMCH in the zero-pivot check --- it's really 
a check for zero division in the LAPACK code. cf e.g.
https://github.com/Reference-LAPACK/lapack/blob/master/SRC/dgetrf2.f#L194
.
DLAMCH however does control the condition number check with INFO > n.

**Comment by ilayn** [comment]
That's interesting. There is another explicit check around line 594-ish for setting the matrix being singular after a call to ?GECON http://www.netlib.org/lapack/explore-html/d7/d3b/group__double_g_esolve_ga7091d0af8d314e611bc190d6f515e7ff.html#ga7091d0af8d314e611bc190d6f515e7ff

**Comment by matthew-brett** [comment]
Any progress here?  I'm hoping to get the scipy daily wheel builds working, and this error is the current blocker.

**Comment by ilayn** [comment]
based on the response you get, I've changed the test to the one that we discussed in the comments. If that also fails, we can fall back to accepting LinAlgError exception
